File: 20-ValidParentheses/20-ValidParentheses.java
// Last updated: 04/10/2025, 01:46:35
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);

            if(ch == '(' || ch == '{' || ch == '['){
                stack.push(ch);
            } else {
                if(stack.isEmpty()) return false;

                if(ch == ')' && stack.peek() == '('){
                    stack.pop();
                } else if(ch == '}' && stack.peek() == '{'){
                    stack.pop();
                } else if(ch == ']' && stack.peek() == '['){
                    stack.pop();
                } else {
                    return false;
                }
            }
        }

        if(stack.isEmpty()){
            return true;
        } else {
            return false;
        }
    }
}

File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 04/10/2025, 01:46:41
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        mapping = {")": "(", "}": "{", "]": "["}
        
        for char in s:
            if char in mapping:
                # If the current character is a closing bracket
                top_element = stack.pop() if stack else '#'  # Pop the top element from the stack, if not empty
                if mapping[char] != top_element:
                    return False  # If the top element does not match the corresponding opening bracket, return False
            else:
                stack.append(char)  # If the current character is an opening bracket, push it onto the stack
        
        return not stack  # If the stack is empty, all brackets were properly matched, return True; otherwise, return False

